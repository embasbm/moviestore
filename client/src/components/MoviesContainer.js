import React, { Component } from 'react';
import axios from 'axios';
import { Table } from 'react-bootstrap';

import NewMovieForm from './NewMovieForm';
import EditMovieForm from './EditMovieForm';
import Movie from './Movie';
class MoviesContainer extends Component {
  constructor(props) {
    super(props)
    this.state = {
      movies: [],
      editingMovieId: null,
      categories: [],

    }
    this.addNewMovie = this.addNewMovie.bind(this)
    this.removeMovie = this.removeMovie.bind(this)
    this.editingMovie = this.editingMovie.bind(this)
    this.editMovie = this.editMovie.bind(this)
  }

  componentDidMount() {
    axios.get('/api/v1/movies.json')
      .then(response => {
        this.setState({
          movies: response.data
        })
      })
      .catch(error => console.log(error))
    axios.get('/api/v1/categories.json')
      .then(response => {
        this.setState({
          categories: response.data
        })
      })
      .catch(error => console.log(error))
  }

  addNewMovie(title, text, category_id) {
    axios.post('/api/v1/movies.json', { movie: { title, text, category_id } })
      .then(response => {
        const movies = [...this.state.movies, response.data]
        this.setState({ movies })
      })
      .catch(error => {
        console.log(error)
      })
  }

  removeMovie(id) {
    axios.delete('/api/v1/movies/' + id)
      .then(response => {
        const movies = this.state.movies.filter(
          movie => movie.id !== id
        )
        this.setState({ movies })
      })
      .catch(error => console.log(error))
  }

  editingMovie(id) {
    this.setState({
      editingMovieId: id
    })
  }

  editMovie(id, title, text, category_id) {
    axios.put('/api/v1/movies/' + id, {
      movie: {
        title,
        text,
        category_id
      }
    })
      .then(response => {
        const movies = this.state.movies;
        const index = movies.indexOf(movies.find((e) => e.id === id));
        if (index >= 0) {
          movies[index] = response.data
          this.setState(() => ({
            movies,
            editingMovieId: null
          }))
        }
      })
      .catch(error => console.log(error));
  }

  render() {
    return (
      <div className="movies-container">
        <NewMovieForm onNewMovie={this.addNewMovie} categories={this.state.categories} />
        <Table striped condensed>
          <tbody>
            {this.state.movies.map(movie => {
              if (this.state.editingMovieId === movie.id) {
                return (
                  <tr key={movie.id}>
                    <EditMovieForm
                      movie={movie}
                      category={movie.category}
                      editMovie={this.editMovie}
                      categories={this.state.categories}
                    />
                  </tr>
                )
              } else {
                return (
                  <tr key={movie.id}>
                    <Movie
                      movie={movie}
                      category={movie.category}
                      onremoveMovie={this.removeMovie}
                      editingMovie={this.editingMovie}
                    />
                  </tr>
                )
              }
            })}
          </tbody>
        </Table>
      </div>
    )
  }
}

export default MoviesContainer;
