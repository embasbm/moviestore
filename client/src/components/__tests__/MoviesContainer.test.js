import React from 'react';
import axios from "axios";
const ReactTestRenderer = require('react-test-renderer');
import MoviesContainer from '../MoviesContainer';

jest.mock('axios');

describe('MoviesContainer', () => {
  const resp = {
    data: []
  };
  axios.get.mockImplementation(() => Promise.resolve(resp));

  it('Should compare MoviesContainer component with a snapshot', () => {
    const component = ReactTestRenderer.create(<MoviesContainer />);
    const json = component.toJSON();
    expect(json).toMatchSnapshot();
  })
});
