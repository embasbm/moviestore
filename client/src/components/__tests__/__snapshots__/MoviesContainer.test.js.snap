// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MoviesContainer Should compare MoviesContainer component with a snapshot 1`] = `
<div
  className="movies-container"
/>
`;
